### PR TITLE
Allow server admins to define and enforce a password policy (MSC2000).

### PR DESCRIPTION
**Pull request by dklimpel**
PR for: #5112
Code from #5214

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: synapse/config/password.py [hunk 1]
```diff
@@ -48,4 +52,29 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
            # DO NOT CHANGE THIS AFTER INITIAL SETUP!
            #
            #pepper: "EVEN_MORE_SECRET"
+
+           # Define and enforce a password policy. Each parameter is optional, boolean
+           # parameters default to 'false' and integer parameters default to 0.
+           # This is an early implementation of MSC2000.
```
Comment (anoadragon453) [inline synapse/config/password.py]: Not sure if this is still "early". I'd just get rid of that word.

File: synapse/config/password.py [hunk 2]
```diff
@@ -48,4 +52,29 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
...
+           #
+           #policy:
+              # Whether to enforce the password policy.
+              #
+              #enabled: true
```
Comment (anoadragon453) [inline synapse/config/password.py]: Could you add spaces between each item here (see the below config block)? It's currently a bit difficult to read.

File: synapse/config/password.py [hunk 3]
```diff
@@ -31,6 +31,10 @@ def read_config(self, config, **kwargs):
         self.password_localdb_enabled = password_config.get("localdb_enabled", True)
         self.password_pepper = password_config.get("pepper", "")
 
+        # Password policy
+        self.password_policy = password_config.get("policy", {})
+        self.password_policy_enabled = self.password_policy.pop("enabled", False)
```
Comment (anoadragon453) [inline synapse/config/password.py]:
```suggestion
        self.password_policy_enabled = self.password_policy.get("enabled", False)
```

While unlikely, something else may try to may access this later.

File: tests/rest/client/v2_alpha/test_password_policy.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+        1. Password is long enough
+        2. Password contains digit(s)
+        3. Password contains symbol(s)
+        4. Password contains uppercase letter(s)
+        5. Password contains lowercase letter(s)
+
+    Therefore, each test in this test case that tests whether a password triggers the
+    right error code to be returned provides a password good enough to pass the previous
+    steps but not the one it's testing (nor any step that comes after).
```
Comment (anoadragon453) [inline tests/rest/client/v2_alpha/test_password_policy.py]: This comment is quite impenetrable. I'd rewrite it as the following:

For each test below that checks whether a password triggers the right error code, that test provides a password good enough to pass the previous tests, but not the one it is currently testing (nor any test that comes afterward).

File: synapse/config/password.py (hunk 2 above)
Comment (richvdh) [reply synapse/config/password.py]: to add to @anoadragon453 's comments: please remember to explicitly document the defaults.

Comment (richvdh) [reply synapse/config/password.py]: oh I see that the defaults are documented further up; I missed them there. I think they should be documented next to the options in question.

File: synapse/config/password.py [hunk 5]
```diff
@@ -48,4 +52,29 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
            # DO NOT CHANGE THIS AFTER INITIAL SETUP!
            #
            #pepper: "EVEN_MORE_SECRET"
+
+           # Define and enforce a password policy. Each parameter is optional, boolean
+           # parameters default to 'false' and integer parameters default to 0.
+           # This is an early implementation of MSC2000.
+           #
+           #policy:
```
Comment (richvdh) [inline synapse/config/password.py]: https://github.com/matrix-org/synapse/blob/master/docs/code_style.md:

> leave the top-level config option uncommented

File: synapse/config/password.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/config/password.py]: Yeah, also allows for different defaults for different policy types, which may be useful in the future.

**Pull request merged**